add CFNumber.h bindings to number.rs

oopsies49 opened the pull request:
Omitted some constants from the header which did not have a value.

oopsies49 committed:
add CFNumber.h bindings to number.rs

Omitted some constants from the header which did not have a value.
# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
# On branch master
# Your branch is up-to-date with 'origin/master'.
#
# Changes to be committed:
#	modified:   src/lib.rs
#	modified:   src/number.rs
#
# Untracked files:
#	src/.base.rs.swp
#	src/.lib.rs.swp
#	src/.number.rs.swp
#

dcuddeback commented on src/number.rs:
```diff
@@ -1,7 +1,48 @@
 
+#[repr(C)]
+struct __CFBoolean;
+pub type CFBooleanRef = __CFBoolean;
```
Should this be `*const __CFBoolean`? `CFNumber.h` defines `CFBooleanRef` like this:

``` c
typedef const struct CF_BRIDGED_TYPE(NSNumber) __CFBoolean * CFBooleanRef;
```

dcuddeback commented on src/number.rs:
```diff
@@ -1,7 +1,48 @@
+    pub fn CFBooleanGetValue(boolean: CFBooleanRef) -> Boolean;
+}
+
+pub type CFNumberType = CFTypeID;
```
Where does `CFTypeID` come from? The header file I'm looking at defines `CFNumberType` as `CFIndex`:

``` c
typedef CF_ENUM(CFIndex, CFNumberType) {
```

oopsies49 replied:
The values of the enum would still be defined as `CFNumberType` though, correct?

``` rust
pub const kCFNumberSInt8Type: CFNumberType = 1;
```

dcuddeback commented on src/number.rs:
```diff
@@ -1,7 +1,48 @@
 extern "C" {
     pub fn CFNumberGetTypeID() -> CFTypeID;
+    pub fn CFNumberCreate(alloc: CFAllocatorRef, theType: CFNumberType, valuePtr: *const c_void) -> CFNumberRef;
+    pub fn CFNumberGetValue(number: CFNumberRef, theType: CFNumberType, valuePtr: *const c_void) -> CFNumberRef;
```
`valuePtr` should be `*mut c_void`:

``` c
CF_EXPORT
Boolean CFNumberGetValue(CFNumberRef number, CFNumberType theType, void *valuePtr);
```

dcuddeback commented on src/number.rs:
```diff
@@ -1,7 +1,48 @@
+    pub fn CFNumberGetType(number: CFNumberRef) -> CFNumberType;
+    pub fn CFNumberGetByteSize(number: CFNumberRef) -> CFIndex;
+    pub fn CFNumberIsFloatType(number: CFNumberRef) -> Boolean;
+    pub fn CFComparisonResult(number: CFNumberRef, otherNumber: CFNumberRef, context: *const c_void) -> CFComparisonResult;
```
Is this function supposed to be `CFNumberCompare`?

``` c
CF_EXPORT
CFComparisonResult CFNumberCompare(CFNumberRef number, CFNumberRef otherNumber, void *context);
```

dcuddeback commented on src/lib.rs:
```diff
@@ -17,3 +18,4 @@ pub mod data;
 pub mod dictionary;
 pub mod locale;
 pub mod string;
+pub mod number;
```
I'd like to keep the modules alphabetized if possible.

dcuddeback commented:
@oopsies49 Thanks for doing this. I found a couple discrepancies between the definitions in the PR and the header file that I get in Xcode. Just curious, which version of the header file are you using? I'm using the one that comes with the OS X 10.10 framework.

Do  you mind reordering the function in the second `extern "C"` block to match the order in the `CFNumber.h` header file? This is the order I see in the header file:

``` c
CFTypeID CFNumberGetTypeID(void);
CFNumberRef CFNumberCreate(CFAllocatorRef allocator, CFNumberType theType, const void *valuePtr);
CFNumberType CFNumberGetType(CFNumberRef number);
CFIndex CFNumberGetByteSize(CFNumberRef number);
Boolean CFNumberIsFloatType(CFNumberRef number);
Boolean CFNumberGetValue(CFNumberRef number, CFNumberType theType, void *valuePtr);
CFComparisonResult CFNumberCompare(CFNumberRef number, CFNumberRef otherNumber, void *context);
```

oopsies49 committed:
fix code review changes to CFNumber bindings

oopsies49 commented:
Happy to contribute. I'm also on OSX 10.10... not sure how to get a version number of the framework. Based upon your suggested changes, I would say we are on the same version. Just some late night oversights to fix is all.

dcuddeback commented on src/number.rs:
```diff
@@ -1,7 +1,47 @@
+
 extern "C" {
     pub fn CFNumberGetTypeID() -> CFTypeID;
+    pub fn CFNumberCreate(alloc: CFAllocatorRef, theType: CFNumberType, valuePtr: *mut c_void) -> CFNumberRef;
```
Oops. Looks like this got changed from `*const` to `*mut` in your last commit. Should be `*const`.

dcuddeback commented on src/number.rs:
```diff
@@ -1,7 +1,47 @@
+    pub fn CFNumberGetType(number: CFNumberRef) -> CFNumberType;
+    pub fn CFNumberGetByteSize(number: CFNumberRef) -> CFIndex;
+    pub fn CFNumberIsFloatType(number: CFNumberRef) -> Boolean;
+    pub fn CFNumberGetValue(number: CFNumberRef, theType: CFNumberType, valuePtr: *mut c_void) -> CFNumberRef;
```
I overlooked this yesterday. The return type for `CFNumberGetValue` is `Boolean`.

oopsies49 committed:
fix code review of adding CFNumber.h bindings

CFNumberGetValue should return Boolean
CFNumberCreate param valuePtr should be *const
Change I think was missed in code review: CFNumberCompare param context
should be *mut
# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
# On branch master
# Your branch is up-to-date with 'origin/master'.
#
# Changes to be committed:
#	modified:   src/number.rs
#

oopsies49 commented:
One thing I noticed when I was double checking things: I think `CFNumberCompare` parameter context should be `*mut`. Nothing in the header indicates that it is used for anything, it is not indicated to be `const`.

dcuddeback commented:
@oopsies49 Yeah, you're right. Good catch. The documentation just says "Pass `NULL`," so I assume it's not used. I agree that it should be `*mut` to match the C prototype.

dcuddeback commented:
@oopsies49 The documentation that I'm referring to is here: https://developer.apple.com/library/mac/documentation/CoreFoundation/Reference/CFNumberRef/#//apple_ref/c/func/CFNumberCompare

dcuddeback commented:
@oopsies49 Looks good! :+1:

dcuddeback commented:
@oopsies49 Published it to crates.io as v0.1.1.
